File: examples/tree.ml
type tree = Leaf | Node of int * tree * tree

let rec sum_tree t =
  match t with
  | Leaf -> 0
  | Node (v, l, r) -> v + sum_tree l + sum_tree r

let () = print_int (sum_tree (Node (2, Node (1, Leaf, Leaf), Node (3, Leaf, Leaf))))

File: examples/tree.erl
-module(tree).
-export([main/0]).

%% A tiny binary tree ADT.
%% Constructors:  leaf            (nullary)
%%                {node, V, L, R} (int value + two subtrees)

sum_tree(leaf)            -> 0;
sum_tree({node, V, L, R}) -> V + sum_tree(L) + sum_tree(R).

main() ->
    sum_tree({node, 2,
                {node, 1, leaf, leaf},
                {node, 3, leaf, leaf}}).
